File: src-architecture/game/creator/SystemCreator.java
package game.creator;

import game.architecture.GameComponentContainer;
import model.celestials.*;
import model.celestials.parts.Clouds;
import model.celestials.parts.Orbit;
import model.celestials.parts.Ring;
import model.celestials.parts.Sphere;
import model.system.parts.SkyBox;

import javax.media.opengl.GL2;
import java.io.IOException;
import java.util.List;

import static game.architecture.Constants.*;

/**
 * Class containing utility methods used to create planetary systems.
 * <p>
 * TODO systems should be parsed from JSON files.
 */
public abstract class SystemCreator {

    /* ========== PROTECTED ========== */
    protected static SkyBox createSkyBox(String name) {
        return new SkyBox(new Sphere(name, 110, 0.0, 0));
    }

    protected static CelestialBody planet(GL2 gl, String name, String cloudsFileName, String ringName, double radius,
                                          double rotationSpeed, double inclination, CelestialBody orbittingBody, double orbitRadius,
                                          double orbitSpeed, double positionInOrbit, List<CelestialBody> bodies) throws IOException {

        radius *= RADIUS_PARAMETER;
        orbitRadius *= AU_PARAMETER;
        orbitSpeed *= ORBITTING_PARAMETER;

        Ring ring = createRing(ringName, radius);
        Clouds clouds = createClouds(cloudsFileName, radius);

        Sphere sphere = createSphere(gl, name, radius, rotationSpeed, inclination);
        Orbit orbit = createOrbit(orbittingBody, orbitRadius, orbitSpeed, positionInOrbit);

        return addObject(new Planet(name, orbit, clouds, ring, sphere), bodies);
    }

    protected static CelestialBody wormhole(GL2 gl, String name, double radius, double rotationSpeed,
                                            double inclination, CelestialBody orbittingBody, double orbitRadius, double orbitSpeed,
                                            double positionInOrbit, long systemFromId, long systemToId, List<CelestialBody> bodies) throws IOException {

        radius *= RADIUS_PARAMETER;
        orbitRadius *= AU_PARAMETER;
        orbitSpeed *= ORBITTING_PARAMETER;

        Sphere sphere = createSphere(gl, name, radius, rotationSpeed, inclination);
        Orbit orbit = createOrbit(orbittingBody, orbitRadius, orbitSpeed, positionInOrbit);

        return addObject(new Wormhole(name, orbit, sphere, systemFromId, systemToId), bodies);
    }

    protected static Star star(GL2 gl, String name, double radius, double rotationSpeed, List<CelestialBody> bodies)
            throws IOException {

        radius *= RADIUS_PARAMETER;

        Sphere sphere = createSphere(gl, name, radius, rotationSpeed, 0);
        Orbit orbit = new Orbit(null, 0, 0, 0);

        return (Star) addObject(new Star(name, orbit, sphere), bodies);
    }

    protected static CelestialBody moon(GL2 gl, String name, String cloudsFileName, double radius,
                                        double rotationSpeed, double inclination, CelestialBody orbittingBody, double orbitRadius,
                                        double orbitSpeed, double positionInOrbit, List<CelestialBody> bodies) throws IOException {

        radius *= RADIUS_PARAMETER;
        orbitRadius *= AU_PARAMETER * MOON_VARIABLE;
        orbitSpeed *= ORBITTING_PARAMETER;

        Sphere sphere = createSphere(gl, name, radius, rotationSpeed, inclination);
        Orbit orbit = createOrbit(orbittingBody, orbitRadius, orbitSpeed, positionInOrbit);

        return addObject(new Moon(name, orbit, sphere), bodies);
    }

    private static CelestialBody addObject(CelestialBody body, List<CelestialBody> bodies) {
        bodies.add(body);
        GameComponentContainer.celestialBodyRepository.add(body);

        return body;
    }

    private static Clouds createClouds(String file, double radius) {
        return file != null ? new Clouds(file, radius + CLOUDS_RADIUS) : null;
    }

    private static Ring createRing(String file, double radius) {
        return file != null ? new Ring(file, radius + INNER_RING_RADIUS, radius + OUTER_RING_RADIUS) : null;
    }

    private static Sphere createSphere(GL2 gl, String name, double radius,
                                       double rotationSpeed, double inclination) {

        return new Sphere(name, radius, rotationSpeed * ROTATION_PARAMETER, inclination);
    }

    private static Orbit createOrbit(CelestialBody orbittingBody, double orbitRadius, double orbitSpeed,
                                     double positionInOrbit) {

        return new Orbit(orbittingBody, orbitRadius, orbitSpeed, positionInOrbit);
    }
}

File: src-architecture/game/creator/SystemCreator.kt
package game.creator

import engine.math.Degree
import engine.math.Radius
import engine.math.ScaleUnit
import engine.math.Velocity
import game.architecture.Constants
import game.architecture.Constants.*
import game.architecture.GameComponentContainer
import model.celestials.CelestialBody
import model.celestials.Planet
import model.celestials.Star
import model.celestials.Wormhole
import model.celestials.parts.*
import model.system.parts.SkyBox

abstract class SystemCreator {

    fun createSkyBox(name: String): SkyBox {
        return SkyBox(
                Sphere(
                        name,
                        Radius(110f),
                        Velocity(0.0f),
                        Degree.ZERO))
    }

    fun planet(
            name: String,
            cloudsFileName: String?,
            ringName: String?,
            radius: Float,
            rotationSpeed: Float,
            inclination: Double,
            orbitingBody: CelestialBody,
            orbitRadius: Float,
            orbitSpeed: Float,
            positionInOrbit: Float,
            bodies: MutableList<CelestialBody>
    ): CelestialBody {

        val ring = createRing(ringName, radius)
        val clouds = createClouds(cloudsFileName, radius)

        val sphere = Sphere(
                name,
                Radius(radius),
                Velocity(rotationSpeed),
                Degree(inclination))

        val orbit = createOrbit(
                orbitingBody,
                Velocity(Constants.ORBITTING_PARAMETER * orbitSpeed),
                positionInOrbit,
                Radius(orbitRadius, ScaleUnit.AU))

        return addObject(Planet(name, orbit, clouds, ring, sphere), bodies)
    }

    fun moon(
            name: String,
            radius: Float,
            rotationSpeed: Float,
            inclination: Double,
            orbitingBody: CelestialBody,
            orbitRadius: Float,
            orbitSpeed: Float,
            positionInOrbit: Float,
            bodies: MutableList<CelestialBody>
    ): CelestialBody {
        return planet(
                name,
                null,
                null,
                radius * 1.5f,
                rotationSpeed,
                inclination,
                orbitingBody,
                // TODO moons do not work well
                orbitRadius,
                orbitSpeed,
                positionInOrbit,
                bodies)
    }

    fun star(
            name: String,
            radius: Float,
            rotationSpeed: Float,
            bodies: MutableList<CelestialBody>
    ): Star {

        val sphere = Sphere(
                name,
                Radius(radius),
                Velocity(rotationSpeed),
                Degree.ZERO)

        val orbit = createOrbit(null, Velocity(0f), 0f, Radius(0f, ScaleUnit.AU))

        return addObject(Star(name, orbit, sphere), bodies) as Star
    }

    fun wormhole(
            name: String,
            radius: Float,
            rotationSpeed: Float,
            inclination: Double,
            orbittingBody: CelestialBody,
            orbitRadius: Float,
            orbitSpeed: Float,
            positionInOrbit: Float,
            systemFromId: Long,
            systemToId: Long,
            bodies: MutableList<CelestialBody>): CelestialBody {

        val sphere = Sphere(
                name,
                Radius(radius),
                Velocity(rotationSpeed),
                Degree(inclination))

        val orbit = createOrbit(
                orbittingBody,
                Velocity(orbitSpeed),
                positionInOrbit,
                Radius(orbitRadius, ScaleUnit.AU))

        return addObject(Wormhole(name, orbit, sphere, systemFromId, systemToId), bodies)
    }

    private fun addObject(body: CelestialBody, bodies: MutableList<CelestialBody>): CelestialBody {
        bodies.add(body)
        GameComponentContainer.celestialBodyRepository.add(body)
        return body
    }

    private fun createClouds(textureFile: String?, radius: Float): Clouds {
        return if (textureFile != null) {
            val cloudsRadius = Radius(radius + CLOUDS_RADIUS)
            Clouds(textureFile, cloudsRadius)
        } else {
            EmptyClouds()
        }
    }

    private fun createRing(textureFile: String?, radius: Float): Ring {
        return if (textureFile != null) {
            val innerRadius = Radius(radius + INNER_RING_RADIUS)
            val outerRadius = Radius(radius + OUTER_RING_RADIUS)
            Ring(textureFile, innerRadius, outerRadius)
        } else {
            EmptyRing()
        }
    }

    private fun createOrbit(
            orbittingBody: CelestialBody?,
            orbitSpeed: Velocity,
            positionInOrbit: Float,
            radius: Radius
    ): Orbit {

        return Orbit(orbittingBody, radius, orbitSpeed, Degree(positionInOrbit))
    }
}